### PR TITLE
Use typed IDs over arena::Index in ssa pass

### PR DESCRIPTION
**Pull request by jfecher**
This PR is meant as a cleanup with no intended behavior changes*.

It:
- Replaces `arena::Index` in favor of `BlockId` and `NodeId` which gives us more type safety. Since we are effectively using indices as pointer, it is similar to replacing `void*` in favor of `BasicBlock*` and `NodeObj*`.
- Renames some parameters. I had no formal way of doing this change, I only changed some parameter names to be more descriptive as I saw them rather than attempting to go through every function. I tried to name all uses of `IRGenerator` as `igen` where before it was split between `igen`, `cfg` and `eval`.

*Refactoring to using stronger types via BlockId and InstructionId did reveal a bug in unrolling for loops where a node index was being passed as a block index.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/integer.rs [hunk 1]
```diff
@@ -324,37 +324,29 @@ pub fn block_overflow(
...
+                ins.rhs = igen.get_or_create_const(trunc_size, node::ObjectType::Unsigned(32));
             }
-            update_ins_parameters(eval, ins.idx, l_new, r_new, max_r_value);
+            update_ins_parameters(igen, ins.id, l_new, r_new, max_r_value);
         }
     }
-    update_value_array(eval, b_idx, &value_map);
-    let mut anchor: HashMap<node::Operation, VecDeque<arena::Index>> = HashMap::new();
+    update_value_array(igen, block_id, &value_map);
+    // let mut anchor = HashMap::new();
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa/integer.rs:333]: dead comment

File: crates/noirc_evaluator/src/ssa/node.rs [hunk 2]
```diff
@@ -281,15 +301,16 @@ impl NodeEval {
...
-        rhs: arena::Index,
+        lhs: NodeId,
+        rhs: NodeId,
         r_type: ObjectType,
-        parent_block: Option<arena::Index>,
+        parent_block: Option<BlockId>,
     ) -> Instruction {
-        let id0 = crate::ssa::code_gen::IRGenerator::dummy_id();
-        let p_block = parent_block.unwrap_or(id0);
+        let id = NodeId(IRGenerator::dummy_id());
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa/node.rs:309]: We can use `NodeId::dummy()` here for `id`

**Review by kevaundray — APPROVED** [review]

File: crates/noirc_evaluator/src/ssa/integer.rs [hunk 3]
```diff
@@ -324,37 +324,29 @@ pub fn block_overflow(
...
+
     //We run another round of CSE for the block in order to remove possible duplicated truncates, this will assign 'new_list' to the block instructions
-    optim::block_cse(eval, b_idx, &mut anchor, &mut new_list);
+    // TODO: Type error, new_list: Vec<NodeId>, but block_cse expects blocks: Vec<BlockId>
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply crates/noirc_evaluator/src/ssa/integer.rs:336]: Guillaume mentioned that this is fine since we do not use jumps yet

[the remark this answers is not in the document]
Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/integer.rs]: This is a different case than was mentioned in the PR. The other type error was with unrolling for loops and I've fixed it by passing the first instruction of the block instead of the block itself. It is good evidence this approach lets us catch bugs though :)

The case from this comment was actually from a bad translation on my part where block_cse had a parameter `block_list` that I thought was a list of blocks, where it should have actually been a list of instructions. I fixed the naming issue and type but forgot to uncomment this part.

File: crates/noirc_evaluator/src/ssa/node.rs [hunk 4]
```diff
@@ -260,19 +279,20 @@ impl std::fmt::Display for Instruction {
 #[derive(Debug, Clone, Copy)]
 pub enum NodeEval {
     Const(FieldElement, ObjectType),
-    Idx(arena::Index),
+    Instruction(NodeId),
```
Comment (guipublic) [inline crates/noirc_evaluator/src/ssa/node.rs:282]: Why did you rename Idx into Instruction?
AFAIK, the underlying NodeId is not necessarily an instruction, it can also be a Variable.

File: crates/noirc_evaluator/src/ssa/ssa_form.rs [hunk 5]
```diff
@@ -70,40 +63,32 @@ pub fn get_block_value(
...
-        return val;
-    }
-    //Global value numbering
-    get_block_value(igen, root, block_id)
+    var_id: NodeId,
+    block_id: BlockId,
+) -> NodeId {
+    let root = igen.get_root_value(var_id);
+
+    igen[block_id]
```
Comment (guipublic) [inline crates/noirc_evaluator/src/ssa/ssa_form.rs:86]: Can we keep the comments?
```
 igen[block_id]
 .get_current_value(root)                                                              //Local value numbering
        .unwrap_or_else(|| get_block_value(igen, root, block_id))    //Global value numbering
```

**Review by guipublic — CHANGES_REQUESTED** [review]

File: crates/noirc_evaluator/src/ssa/node.rs (hunk 4 above)
Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/node.rs:282]: Ah, I'll change it to something more explicit like `VarOrInstruction` in that case

**Review by kevaundray — APPROVED** [review]

**Review by kevaundray — APPROVED** [review]

**Review by guipublic — APPROVED** [review]
